torch/distributed/tensor/_dtensor_spec.py

```diff
@@ -57,17 +106,16 @@ def _hash_impl(self) -> int:
         # dtype and stride.
         # Caveat: we need to keep this in mind and sync hash and eq if we add more
         # fields to them.
+        hash_items = [self.mesh, self.placements, self.shard_order]
         if self.tensor_meta is not None:
-            return hash(
-                (
-                    self.mesh,
-                    self.placements,
+            hash_items.extend(
+                [
                     self.tensor_meta.shape,
                     self.tensor_meta.stride,
                     self.tensor_meta.dtype,
-                )
+                ]
             )
-        return hash((self.mesh, self.placements))
+        return hash(tuple(hash_items))
```

Comment:
This is hot path. Better to just deal with the code duplication here, it should result in less allocs. cc @swolchok